polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,291 @@
+                )
+                seen.update(uprns)
+
+
+class DistrictSet(CustomSet, AssignPollingStationsMixin):
+    def build_namedtuple(self, element):
+
+        # MultiPolygon is mutable, so we must serialize it to store in a tuple
+        area = element["area"].ewkb  # use ewkb so it encodes srid
+
+        return District(
+            element.get("name", ""),
+            element["council"],
+            element["internal_council_id"],
+            element.get("extra_id", ""),
+            area,
+            element.get("polling_station_id", ""),
+        )
+
+    def save(self):
+        districts_db = []
+        for district in self.elements:
+            record = PollingDistrict(
+                name=district.name,
+                council=district.council,
+                internal_council_id=district.internal_council_id,
+                extra_id=district.extra_id,
+                area=district.area,
+                polling_station_id=district.polling_station_id,
+            )
+            districts_db.append(record)
+        PollingDistrict.objects.bulk_create(districts_db)
+
+    def get_polling_station_lookup(self):
+        """
+        for each address, build a lookup of polling_station_id -> set of uprns
+        This does the lookup in the db so you need to have called save() before
+        using this.
+        """
+        polling_station_lookup = {}
+        cursor = connection.cursor()
+        cursor.execute(
+            """ 
+                SELECT a.uprn, d.polling_station_id
+                FROM addressbase_address a
+                    JOIN addressbase_uprntocouncil u
+                    ON a.uprn = u.uprn
+                    JOIN pollingstations_pollingdistrict d
+                    ON ST_Contains(d.area, a.location)
+                WHERE d.council_id=%s
+                AND u.lad=%s
+            """,
+            [self.council_id, self.council_id],
+        )
+        for row in cursor.fetchall():
+            uprn = row[0]
+            polling_station_id = row[1]
+            if polling_station_id not in polling_station_lookup:
+                polling_station_lookup[polling_station_id] = {uprn}
+            else:
+                polling_station_lookup[polling_station_id].add(uprn)
+
+        return polling_station_lookup
+
+
+class StationSet(CustomSet, AssignPollingStationsMixin):
+    def build_namedtuple(self, element):
+
+        # Point is mutable, so we must serialize it to store in a tuple
+        if "location" in element and element["location"]:
+            location = element["location"].ewkb  # use ewkb so it encodes srid
+        else:
+            location = None
+
+        return Station(
+            element["council"],
+            element["internal_council_id"],
+            element.get("postcode", ""),
+            element.get("address", ""),
+            location,
+            element.get("polling_district_id", ""),
+        )
+
+    def get_polling_station_lookup(self):
```

Comment:
I'm confused about when we call `StationSet.get_polling_station_lookup()` instead of `DistrictSet.get_polling_station_lookup()` or `AddressList.get_polling_station_lookup()`. I think I can see what it is supposed to do and where I think we should go with this, but the only place I can see we're calling it is in unit tests and `update_uprn_to_council_model()` and then as far as I can tell, we only ever call `DistrictSet.update_uprn_to_council_model()` or `AddressList.update_uprn_to_council_model()` in `base_importers.py`. As a fair step, can you explain what this method does and when/where we call it?

Reply:
29c81e6

Reply:
Yes, this has uncovered something I've only half done - thank you!

The intention was to cover the fact that when we import districts and stations (as opposed to addresses and stations) we don't know if the FK will be in the districts dataset (i.e. a stations id) or the stations dataset (i.e. a district id). Having  `StationSet.get_polling_station_lookup()` is intended to cover the second case. In more words: provide a 'polling station -> set of uprns' lookup derived from uprn location falling within a district, where the district doesn't have a polling station id, but the polling stations do have a district id. 

As you've pointed out it never gets called - and figuring out what station to return to the user still happens in `PollingStationManager.get_polling_station()` which is called from `data_finder.views` and in the api. My intention had been that it would always come from the `UprnToCouncil` table which is essentially caching the point in polygon lookup that `PollingStationManager` does.

I guess this is where you'd got to already :slightly_smiling_face: . Do you think this is generally a good direction to be going in? 

If I did implement it I think I'd do the following:
  - Modify `BaseStationsDistrictsImporter.import_data()` to check which table had id's filled in and then call `update_uprn_to_council_model()` from the appropriate `{District,Station}Set`. Something like:

    ```python
        def districts_have_station_ids(self):
            # Go and check

        def import_data(self):
            # ...no change
            self.import_polling_districts()
            self.import_polling_stations()
            self.districts.save()
            self.stations.save()
            if self.districts_have_station_ids():
                self.districts.update_uprn_to_council_model()
            else:
                self.stations.update_uprn_to_council_model()     
    ```
  - change the  `{Address,Postcode}View.get_station()` methods to look in the `UprnToCouncil` table. 
  - change the API `generate_polling_station` methods similarly. 

The second and third steps don't seem that satisfactory as they don't simplify the routing story very much and would duplicate the lookups done in `routing.py` to determine which view to return - but maybe no worse than repeating the point in polygon lookup - I doubt either are that expensive anyway.

Writing the above has made me realise that I think `routing.py` needs the changes to `BaseStationsDistrictsImporter.import_data()` anyway - so maybe that's all I should do in this PR and figure out the routing story in a smaller chunk... :thinking:

Reply:
Yeah I figured that was the case it was supposed to cover and that it has been overlooked :)

I think you are totally right to say we should always work out the mapping of addresses --> stations here and write the values to `UprnToCouncil`. Then `PollingStationManager.get_polling_station()` _only_ looks at the `UprnToCouncil` table. Essentially the goal here should be that regardless of whether the source data was a list of addresses or district shapes, the output of that is we've now got Polling Station IDs attached to UPRNs in the DB. Tbh, although its probably useful to have the district shapes in the DB for checking, once we've done that we should be able to delete the district shapes completely and it changes nothing. I haven't looked at the `data_finder` or API code all all yet (I will get to it, but I haven't yet), so I'm a bit confused about those comments at this stage, but this should simplify everything because this should move us to a situation where there is only one way to take an input postcode/address and work out a polling station (or lack thereof) and it is the same as if the source data was addresses.

In terms of the code structure for the importers, I don't think it makes sense for this method to live on `StationSet` because although this involves interacting with the stations table, doing this only makes sense if we have districts (if we've got an `AddressList` and a `StationSet`, calling `StationSet.get_polling_station_lookup()` would make no sense). Given this only makes sense if we have districts, it is probably better to think of this as:


`data_types.py`

```py
class DistrictSet(CustomSet, AssignPollingStationsMixin):

    def get_polling_station_lookup(self, districts_have_station_ids):
        if not self.saved:
            raise Exception('oh noes!!')
        polling_station_lookup = {}
        cursor = connection.cursor()
	if districts_have_station_ids:
	    sql = "One somplicated SQL statement <snip>"
        else:
	    sql = "A different complicated SQL statement <snip>"
        cursor.execute(sql)
        for row in cursor.fetchall():
            uprn = row[0]
            polling_station_id = row[1]
            if polling_station_id not in polling_station_lookup:
                polling_station_lookup[polling_station_id] = {uprn}
            else:
                polling_station_lookup[polling_station_id].add(uprn)

        return polling_station_lookup


class StationSet(CustomSet):  # StationSet doesn't need to implement AssignPollingStationsMixin
    def build_namedtuple(self, element):
        # bla

    def save(self):
        # bla
```

`base_importers.py`

```py
    @property
    def districts_have_station_ids(self):
        # Check which way round the mapping is. If it is at all ambiguous, throw an exception

    def import_data(self):
        # ...no change
        self.import_polling_districts()
        self.import_polling_stations()
        self.districts.save()
        self.stations.save()
        self.districts.update_uprn_to_council_model(self.districts_have_station_ids)
```

Also, I've not looked at the tests yet, but I suspect the fact this got overlooked is an indication that we're missing test coverage for this situation.

Reply:
Yep - that makes sense to have both queries on `DistrictSet`. 

Re. the tests you're right it's not covered. ATM working on the tests felt pretty tricky in terms of editing all the fixtures, and getting them all set up to simulate each case - I've tried to move towards one fixture per scenario to make it easier to reason about. However I've not done much to extend the coverage, beyond what was there. I've been wondering if it's worth moving to factories similar to EveryElection before trying to extend the coverage.